aptos-move/framework/aptos-token/sources/bid.move

```diff
@@ -0,0 +1,343 @@
+        token_id: TokenId,
+        token_amount:u64,
+        offer_price: u64,
+        entry: &Listing<CoinType>,
+    ): ID acquires BidRecords {
+        initialize_bid_records<CoinType>(bidder);
+        let bidder_address = signer::address_of(bidder);
+        // check the bid is legit for the listing
+        assert_bid_for_a_listing(token_id, offer_price * token_amount, token_amount, entry);
+        // check bidder has sufficient balance
+        assert!(coin::balance<CoinType>(bidder_address) >= offer_price * token_amount, error::invalid_argument(ENO_SUFFICIENT_FUND));
+        // withdraw the coin and store them in escrow to ensure the fund is avaliable until expiration_sec
+        let coin = coin::withdraw<CoinType>(bidder, offer_price * token_amount);
+
+        let bid = Bid<CoinType> {
+            bidder: bidder_address,
+            coin,
+            offer_price,
+            listing_id: listing::get_listing_id(entry),
+        };
+        let bid_id = create_bid_id(bidder);
+        initialize_bid_records<CoinType>(bidder);
+        let bid_records = borrow_global_mut<BidRecords<CoinType>>(bidder_address);
+        table::add(&mut bid_records.records, bid_id, bid);
+        event::emit_event<BidEvent<CoinType>>(
+            &mut bid_records.bid_event,
+            BidEvent<CoinType> {
+                offer_price,
+                listing_id: listing::get_listing_id(entry),
+                bid_id,
+            },
+        );
+
+        bid_id
+    }
+
+    /// bidder can withdraw fund from any bid when lock time ends
+    public entry fun withdraw_fund_from_bid<CoinType>(
+        bidder: &signer,
+        bid_id_creation_number: u64,
+    ) acquires BidRecords {
+        let bidder_address = signer::address_of(bidder);
+        let bid_id = guid::create_id(bidder_address, bid_id_creation_number);
+
+        let bid_records = borrow_global_mut<BidRecords<CoinType>>(bidder_address);
+        assert!(table::contains(&bid_records.records, bid_id), EBID_NOT_EXIST);
+
+        let bid = table::remove(&mut bid_records.records, bid_id);
+        coin::deposit(bidder_address, clear_bid(bid));
+        event::emit_event<WithdrawBidEvent<CoinType>>(
+            &mut bid_records.withdraw_bid_event,
+            WithdrawBidEvent<CoinType> {
+                bid_id
+            },
+        );
+    }
+
+    fun deduct_fee<CoinType>(
+        total_coin: &mut Coin<CoinType>,
+        fee_numerator: u64,
+        fee_denominator: u64
+    ): Coin<CoinType> {
+        let value = coin::value(total_coin);
+        let fee = if (fee_denominator == 0) {
+            0
+        } else {
+            value * fee_numerator/ fee_denominator
+        };
+        coin::extract(total_coin, fee)
+    }
+
+    /// execute a bid to a listing, no signer required to perform this function
+    /// pay fee to 3 party based on a percentage
+    /// deduct royalty and send to the payee account
+    public fun execute_listing_bid<CoinType>(
```

Comment:
This looks like something that should be defined in an auction_house.move and called `complete_auction`. It Also seems like a backdoor, where I can take any successful bid and complete it, even if it isn't the highest bid for an auction.

Reply:
This is a library function supposed to be called by an auction house. It is designed to be flexible to match a listing with a bid chosen the by auction house. 

To execute this function, it needs to have `Listing<CoinType>`, which is owned by auctioneer, who control the auction logic anyway. No anyone can execute this function.

Reply:
Point being that the auctioneer can cheat with the logic as it stands.

Reply:
I don't think this is an issue that we should police. We provide a library and example. If they want to use the library and create any malicious module, let the them shoot on their own feet then?